drivers/mlx90393/mlx90393.c

```diff
@@ -0,0 +1,625 @@
+    /* check availability of the sensor */
+    if ((error = _is_avaiable(dev)) != 0) {
+        _release(dev);
+        DEBUG("[mlx90393] error: device not available\n");
+        return error;
+    }
+    /* store ref temp in dev */
+    if ((error = _read_register(dev, MLX90393_REG_REF_TEMP, &DEV_REF_TEMP)) != 0) {
+        _release(dev);
+        return error;
+    }
+    /* check oversampling and digital filter configuration */
+    if ((DEV_OSR_MAG == 0 && DEV_DIG_FILT == 0) ||
+        (DEV_OSR_MAG == 0 && DEV_DIG_FILT == 1) ||
+        (DEV_OSR_MAG == 1 && DEV_DIG_FILT == 0)) {
+        _release(dev);
+        DEBUG("[mlx90393] error: the configuration of oversampling and digital filter \
+            is not permitted\n");
+        return -EINVAL;
+    }
+    /* magnetic sensor oversampling */
+    if ((error = _write_register_bits(dev, MLX90393_REG_CONF2, MLX90393_MASK_OSR,
+        DEV_OSR_MAG)) != 0) {
+        _release(dev);
+        return error;
+    }
+    /* magnetic sensor digital filter */
+    uint16_t dig_filt_value_mask = DEV_DIG_FILT << MLX90393_SHIFT_DIG_FILT;
+    if ((error = _write_register_bits(dev, MLX90393_REG_CONF2, MLX90393_MASK_DIG_FILT,
+        dig_filt_value_mask)) != 0) {
+        _release(dev);
+        return error;
+    }
+    /* temperature oversampling */
+    uint16_t temp_osr_value_mask = DEV_OSR_TEMP << MLX90393_SHIFT_OSR2;
+    if ((error = _write_register_bits(dev, MLX90393_REG_CONF2, MLX90393_MASK_OSR2,
+        temp_osr_value_mask)) != 0) {
+        _release(dev);
+        return error;
+    }
+    /* gain */
+    uint16_t gain_value_mask = DEV_GAIN << MLX90393_SHIFT_GAIN;
+    if ((error = _write_register_bits(dev, MLX90393_REG_CONF0, MLX90393_MASK_GAIN_SEL,
+        gain_value_mask)) != 0) {
+        _release(dev);
+        return error;
+    }
+    /* resolution */
+    uint16_t xyz_resolution_value_mask = (DEV_RESOLUTION << MLX90393_SHIFT_RES_Z) |
+                                         (DEV_RESOLUTION << MLX90393_SHIFT_RES_Y) |
+                                         (DEV_RESOLUTION << MLX90393_SHIFT_RES_X);
+    if ((error = _write_register_bits(dev, MLX90393_REG_CONF2, MLX90393_MASK_RES_XYZ,
+        xyz_resolution_value_mask)) != 0) {
+        _release(dev);
+        return error;
+    }
+    /* burst more */
+    if (DEV_MODE == MLX90393_MODE_BURST) {
+        /* set burst data rate */
+        if ((error = _write_register_bits(dev, MLX90393_REG_CONF1, MLX90393_MASK_BDR,
+            DEV_ODR)) != 0) {
+            _release(dev);
+            return error;
+        }
+        /* start burst mode */
+        if ((error = _write_byte(dev, MLX90393_COMMAND_SB)) != 0) {
+            _release(dev);
+            return error;
+        }
+        if ((error = _check_status_byte(dev)) != 0) {
+            _release(dev);
+            return error;
+        }
+    }
+    /* wake up on change mode */
+    else if (DEV_MODE == MLX90393_MODE_WOC_ABSOLUTE || DEV_MODE == MLX90393_MODE_WOC_RELATIVE) {
```

Comment:
Ich überlege noch, ob man den Code weiter optimieren könnte, wenn man ein Pseudomodule `mlx90393_int` einführt, mit dem die Interrupt-Nutzung aktiviert wird.

Erstens würde das den kompletten `else`-Zweig heraus optimieren, wenn das Module nicht genutzt wird
```suggestion
    else if (IS_USED(MODULE_MLX90393_INT) &&
             (DEV_MODE == MLX90393_MODE_WOC_ABSOLUTE || DEV_MODE == MLX90393_MODE_WOC_RELATIVE)) {
```
was nicht passiert, wenn nur auf `dev->mode` getestet wird, weil zur Compile-Zeit nicht klar ist, welchen Wert `dev->mode` hat.

Zweitens könnte man den Member `mlx90393_params_t::int_pin` sparen, wenn es nicht benutzt wird bzw. den Member `mlx90393_t::conversion_time` sowie den gesamten Code mit Berechnung und Nutzung von `_calculate_conv_time` heraus optimieren, wenn `mlx90393_irq` benutzt wird.

Wenn `mlx90393_int` enabled ist, reicht ein einfaches `assert(gpio_is_valid(params->int_pin))` um sicherzustellen, dass das Interrupt Pin für die Interrupt-Nutzung definiert ist, was wiederum Code spart.